### PR TITLE
TSK-1723: Synchronization of LDAP data for users

### PR DESCRIPTION
**Pull request by gitgoodjhe**
---
 
### For the submitter:
- [x] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [ ] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [x] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

**Comment by gitgoodjhe** [comment]
Sonarcloud: https://sonarcloud.io/dashboard?branch=TSK-1723&id=gitgoodjhe_taskana

Updated documentation: https://taskana.atlassian.net/wiki/spaces/TAS/pages/687308819/Job+Configuration

updated release notes: https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana

File: lib/taskana-core/src/main/java/pro/taskana/TaskanaEngineConfiguration.java [hunk 1]
```diff
@@ -62,7 +62,11 @@
   private static final String TASKANA_JOB_PRIORITY_BATCHSIZE = "taskana.jobs.priority.batchSize";
   private static final String TASKANA_JOB_PRIORITY_RUN_EVERY = "taskana.jobs.priority.runEvery";
   private static final String TASKANA_JOB_PRIORITY_FIRST_RUN = "taskana.jobs.priority.firstRunAt";
+  private static final String TASKANA_JOB_USER_REFRESH_FIRST_RUN =
+      "taskana.jobs.user.refresh.firstRunAt";
   private static final String TASKANA_JOB_PRIORITY_ACTIVE = "taskana.jobs.priority.active";
+  private static final String TASKANA_JOB_USER_REFRESH_RUN_EVERY =
```
Comment (holgerhagen) [inline lib/taskana-core/src/main/java/pro/taskana/TaskanaEngineConfiguration.java:68]: sort variables?

File: rest/taskana-rest-spring-example-common/src/test/java/pro/taskana/example/ldap/LdapTest.java [hunk 2]
```diff
@@ -48,4 +52,40 @@ void should_ReturnFullDnForUser_When_AccessIdOfUserIsGiven() throws Exception {
     String dn = ldapClient.searchDnForAccessId("user-2-2");
     assertThat(dn).isEqualTo("uid=user-2-2,cn=users,ou=test,o=taskana");
   }
+
+  @Test
+  void should_ReturnAllUsersInUserRoleWithCorrectAttributes() {
```
Comment (holgerhagen) [inline rest/taskana-rest-spring-example-common/src/test/java/pro/taskana/example/ldap/LdapTest.java:57]: The configuration contains all individual users in addition to the group. (taskana.roles.user= cn=ksc-users,cn=groups,OU=Test,O=TASKANA | teamlead-1 | teamlead-2 | user-1-1 | user-1-2 | user-2-1 | user-2-2 | user-b-1 | user-b-2)

Are we sure, that the group resolving works correctly? Why do we have this configuration?

File: rest/taskana-rest-spring/src/main/java/pro/taskana/common/rest/ldap/LdapClient.java [hunk 3]
```diff
@@ -128,6 +130,33 @@ public LdapClient(
...
+    Set<String> userGroupsOrUser = taskanaEngineConfiguration.getRoleMap().get(TaskanaRole.USER);
+
+    final OrFilter userOrGroupFilter = new OrFilter();
+    userGroupsOrUser.forEach(
+        userOrGroup -> {
+          userOrGroupFilter.or(new EqualsFilter(getUserMemberOfGroupAttribute(), userOrGroup));
+          userOrGroupFilter.or(new EqualsFilter(getUserIdAttribute(), userOrGroup));
+        });
+
+    final AndFilter andFilter = new AndFilter();
```
Comment (holgerhagen) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/common/rest/ldap/LdapClient.java:144]: Why do we need the "AndFilter"? Can't we use the OrFilter directly?

Comment (holgerhagen) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/common/rest/ldap/LdapClient.java:144]: Or maybe the condition on ldap object type is missing?

File: rest/taskana-rest-spring/src/main/java/pro/taskana/user/jobs/UserInfoRefreshJob.java [hunk 4]
```diff
@@ -0,0 +1,127 @@
+package pro.taskana.user.jobs;
```
Comment (holgerhagen) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/user/jobs/UserInfoRefreshJob.java:1]: In general I would prefer to have some more debug logging in such kind of jobs to support explaining weird behaviour on the customer machine... ;-)

File: rest/taskana-rest-spring/src/test/resources/taskana.properties [hunk 5]
```diff
@@ -1,4 +1,4 @@
-taskana.roles.user=cn=ksc-users,cn=groups,OU=Test,O=TASKANA | teamlead-1 | teamlead-2 | user-1-1 | user-1-2 | user-2-1 | user-2-2 | user-b-1 | user-b-2
+taskana.roles.user=cn=ksc-users,cn=groups,OU=Test,O=TASKANA | teamlead-1 | teamlead-2 | user-1-1
```
Comment (holgerhagen) [inline rest/taskana-rest-spring/src/test/resources/taskana.properties:1]: Is this the answer to my previous comment?

**Review by holgerhagen — CHANGES_REQUESTED** [review]
Do we need a configuration switch to synchronise users during startup (if the user table is empty)?

File: rest/taskana-rest-spring/src/test/resources/taskana.properties (hunk 5 above)
Comment (gitgoodjhe) [reply rest/taskana-rest-spring/src/test/resources/taskana.properties:1]: exactely

File: rest/taskana-rest-spring-example-common/src/test/java/pro/taskana/example/ldap/LdapTest.java (hunk 2 above)
Comment (gitgoodjhe) [reply rest/taskana-rest-spring-example-common/src/test/java/pro/taskana/example/ldap/LdapTest.java:57]: yes, I have taken out some individual users to make sure the group resolve works

**Pull request merged**